### ISSUE TITLE
`@psalm-pure` + `@psalm-assert` declared on `@mixin` referenced interfaces are considered `UnusedMethodCall` when used

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#4381]
The following code is a reduced scenario from what usually happens with a Doctrine ORM + custom DBAL type setup:

```php
<?php

interface Assertions {
    /**
     * @psalm-pure
     * @psalm-assert list $array
     *
     * @param mixed  $array
     */
    static function isList($array): void;

    /**
     * @psalm-pure
     * @psalm-template ExpectedType of object
     * @psalm-param class-string<ExpectedType> $class
     * @psalm-assert iterable<ExpectedType> $value
     *
     * @param mixed $value
     */
    static  function allIsInstanceOf($value, $class): void;
}

/** @mixin Assertions */
class Assert {}

class MyObject
{
    public string $productType = 'foo';
}

/** 
 * @param mixed $value
 * @return string
 */
function convertToDatabaseValue($value): string
{
    Assert::isList($value);
    Assert::allIsInstanceOf($value, MyObject::class);
    assert(\array_key_exists(0, $value));
    
    return $value[0]->productType;
}
```

https://psalm.dev/r/ece65ca3ab

This reports:

```
ERROR: UnusedMethodCall - 37:13 - The call to Assertions::isList is not used

ERROR: UnusedMethodCall - 38:13 - The call to Assertions::allIsInstanceOf is not used
```

Note that the types are correctly inferred based on `@psalm-assert`, so that bit is definitely working!

It seems like some information about `@psalm-assert` + `@psalm-pure` (#2456) is lost when using `@mixin`.

For example, if I move the assertions from the `@mixin Assertions` to the concrete `Assert` class, the `UnusedMethodCall` issues disappear:

```php
<?php

class Assert {
    /**
     * @psalm-pure
     * @psalm-assert list $array
     *
     * @param mixed  $array
     */
    static function isList($array): void {}

    /**
     * @psalm-pure
     * @psalm-template ExpectedType of object
     * @psalm-param class-string<ExpectedType> $class
     * @psalm-assert iterable<ExpectedType> $value
     *
     * @param mixed $value
     */
    static function allIsInstanceOf($value, $class): void {}
}

class MyObject
{
    public string $productType = 'foo';
}

/** 
 * @param mixed $value
 * @return string
 */
function convertToDatabaseValue($value): string
{
    Assert::isList($value);
    Assert::allIsInstanceOf($value, MyObject::class);
    assert(\array_key_exists(0, $value));
    
    return $value[0]->productType;
}
```

https://psalm.dev/r/387bf7c3cf

Also here, I verified that the assertions are 100% working.

Potentially related to #3922, but with a more precise analysis (`@mixin` involved, like #3701)

This was previously fixed for #2456 in https://github.com/vimeo/psalm/commit/4b715cdbffea19a7eab6f72482027d2dd358aab2, but the information about `@mixin` is somehow lost somewhere :)

**Comment by ondrejmirtes** [comment]
FWIW PHPStan doesn't report "Method call on a separate line has no effect" for any pure method that also have `@throws`. Maybe Psalm has a different logic that breaks here.

**Comment by Ocramius** [comment]
@ondrejmirtes yeah, from what I inspected, the methods on an `@mixin` are not really the same as methods on a concrete implementation.

From my end, I'll move away from `@mixin` for this specific use-case

**Comment by Ocramius** [comment]
Closing here: as per @muglug's suggestions, I've improved `webmozart/assert` via https://github.com/webmozarts/assert/pull/214 to get rid of `@mixin` hacks, which reduces complexity overall :-)
